[WIP] Include incompatible dependencies in `outdated` output

mattprowse opened the pull request:
Proposed fix for #1260.

mdiep commented on Source/CarthageKit/Project.swift:
```diff
@@ -353,34 +353,77 @@ public final class Project {
+						return dependency.project
+					}
+			}
+			.flatten(.Merge)
```
This could use `flatMap` instead of `map` and `flatten`.

mdiep replied:
Should there be a `uniqueValues()` here or something to prevent a project from getting sent twice?

(If `A` and `B` both depend on `C`, I think it will get sent twice?)

mdiep commented on Source/CarthageKit/Project.swift:
```diff
@@ -353,34 +353,77 @@ public final class Project {
+					}
+					.ignoreNil()
+			}
+			.flatten(.Merge)
```
Same here.

mdiep commented on Source/CarthageKit/Project.swift:
```diff
@@ -353,34 +353,77 @@ public final class Project {
+		return loadResolvedCartfile()
+			.map {
+				return SignalProducer<Dependency<PinnedVersion>, CarthageError>(values: $0.dependencies)
+					.map {dependency in
```
There should be a space after the `{` here.

mdiep commented on Source/CarthageKit/Project.swift:
```diff
@@ -353,34 +353,77 @@ public final class Project {
+				}
+				
+				return versions
+					.map {(versions: [SemanticVersion]) -> Dependency<SemanticVersion>? in
```
Here too.

mdiep commented on Source/CarthageKit/Project.swift:
```diff
@@ -353,34 +353,77 @@ public final class Project {
+			}
+			.flatten(.Merge)
+			.map { (projectIdentifier: ProjectIdentifier) -> SignalProducer<Dependency<SemanticVersion>, CarthageError> in
+				let versions = self.versionsForProject(projectIdentifier).collect().map { (pinnedVersions: [PinnedVersion]) -> [SemanticVersion] in
```
This should be formatted according to our conventions, where each new SignalProducer operator is on a new line.

```swift
let versions = self.versionsForProject(projectIdentifier)
    .collect()
    .map { (pinnedVersions: [PinnedVersion]) -> [SemanticVersion] in
        ..
    }
```

mdiep commented on Source/CarthageKit/Project.swift:
```diff
@@ -353,34 +353,77 @@ public final class Project {
+							return Dependency(project: projectIdentifier, version: latestVersion)
+						} else {
+							return nil
+						}
```
This `if`/`else` could be replaced with a `map`.

mdiep commented:
I left a few notes. Comment when those are fixed up and I'll take another look.

I'm excited to have this!

mattprowse commented:
Thanks for the feedback!

I've fixed up the formatting issues, switched from using `map()` + `flatten()` to `flatMap()`, and restructured things to get rid of the if/else in `latestDependencies()`. The restructure of `latestDependencies()` also allowed me to get rid of some intermediate variables that were previously required to keep Xcode happy, so that's nice.

Regarding the necessity of `uniqueValues()`, I'm fairly certain this can be left out because the list of dependencies comes from `resolvedCartfile()` which (if I'm understanding things correctly) only returns each dependency once, because at this point the work of resolving dependencies has already been done. I did some quick testing and this seemed to be the case, though I'm obviously not as familiar with Carthage and it's possible that I'm wrong about this. If so, let me know and I'll add it in.

mdiep commented on Source/CarthageKit/Project.swift:
```diff
@@ -353,34 +353,71 @@ public final class Project {
+					.collect()
+					.map { (pinnedVersions: [PinnedVersion]) -> SemanticVersion? in
+						return pinnedVersions
+							.flatMap { SemanticVersion.fromPinnedVersion($0).value }
```
I think it'd be simpler to map to `SemanticVersion`s before `collect()`ing.

```swift
return self.versionsForProject(projectIdentifier)
    .map { SemanticVersion.fromPinnedVersion($0).value }
    .ignoreNil()
    .collect()
    .map { $0.maxElement() }
    .ignoreNil()
    .map { Dependency(project: projectIdentifier, version: $0) }
```

mdiep commented on Source/CarthageKit/Project.swift:
```diff
@@ -353,34 +353,71 @@ public final class Project {
+				
+				var latestDependenciesDictionary = [ProjectIdentifier: SemanticDependency]()
+				for latestDependency in latestDependencies {
+					if let latestVersion = latestDependency.version.pinnedVersion, currentVersion = currentDependenciesDictionary[latestDependency.project] where currentVersion.version != latestVersion {
```
Can you move this `where` clause to the `if` below inside `satisfiableDepndencies.flatMap`? I think that's a more readable approach.

mdiep commented:
Okay, there are two problems that I've discovered:

1. It doesn't work with dependencies that have no tagged versions.

2. It prints results for dependencies that are using a branch when there's nothing new on the branch. (Because it's not on the latest version.)

I think we need to fix those things up before we can release this.

But this is really neat!

mattprowse commented:
Changes made as requested! I also changed the variable name in the loop that builds `currentDependenciesDictionary`, just so that it's consistent with the other two loops in `outdatedDependencies()`.

I'll work on a fix for the two problems you found and comment again once I've got it done!

ikesyo commented:
Closing this in favor of #2254.
